misc/config_tools/launch_config/launch_cfg_gen.py

```diff
@@ -1,269 +1,364 @@
+        params = virtio_block.split(":", maxsplit=1)
+        if len(params) == 1:
+            script.add_virtual_device("virtio-blk", options=virtio_block)
+        else:
+            block_device = params[0]
+            rootfs_img = params[1]
+            var = f"dir_{os.path.basename(block_device)}"
+            script.add_init_command(f"{var}=`mount_partition {block_device}`")
+            script.add_virtual_device("virtio-blk", options=os.path.join(f"${{{var}}}", rootfs_img))
+            script.add_deinit_command(f"unmount_partition ${{{var}}}")
+
+    # Passthrough PCI devices
+    bdf_regex = re.compile("([0-9a-f]{2}):([0-1][0-9a-f]).([0-7])")
+    for passthru_device in vm_launch_etree.xpath(".//passthrough_devices/*/text()"):
+        m = bdf_regex.match(passthru_device)
+        if not m:
+            continue
+        script.add_passthru_device(int(m.group(1), 16), int(m.group(2), 16), int(m.group(3), 16))
+
+    for sriov_gpu_device in vm_launch_etree.xpath(".//sriov/gpu/text()"):
+        m = bdf_regex.match(sriov_gpu_device)
+        if not m:
+            continue
+        script.add_passthru_device(int(m.group(1), 16), int(m.group(2), 16), int(m.group(3), 16), options="igd-vf")
+
+    ###
+    # Miscellaneous
+    ###
+    script.add_dynamic_dm_parameter("add_rtvm_options")
+    script.add_dynamic_dm_parameter("add_logger_settings", "console=4 kmsg=3 disk=5")
+
+    ###
+    # Lastly, conclude the device model parameters with the VM name
+    ###
+    script.add_plain_dm_parameter(f"{vm_name}")
+
+    return script
+
+def main(board_xml, scenario_xml, launch_xml, user_vmid, out_dir):
+    board_etree = etree.parse(board_xml)
+    scenario_etree = etree.parse(scenario_xml)
+    launch_etree = etree.parse(launch_xml)
+
+    sos_vm_id = eval_xpath(scenario_etree, "//vm[vm_type='SERVICE_VM']/@id")
+    post_vms = scenario_etree.xpath("//vm[starts-with(vm_type, 'POST_')]")
+    if sos_vm_id is None and len(post_vms) > 0:
+        logging.error("The scenario does not contain a service VM. Will not generate launch scripts for the post-launched VMs in the scenario.")
+        return 1
+    sos_vm_id = int(sos_vm_id)
+
+    try:
+        os.mkdir(out_dir)
+    except FileExistsError:
+        pass
+    except Exception as e:
+        logging.error(f"Cannot create output directory: {e}")
+        return 1
+
+    if user_vmid == 0:
+        post_vm_ids = [int(vm_scenario_etree.get("id")) - sos_vm_id for vm_scenario_etree in post_vms]
     else:
-        print("Launch files generate failed".format(output))
+        post_vm_ids = [user_vmid]
+
+    for post_vm_id in post_vm_ids:
+        vm_scenario_etree = eval_xpath(scenario_etree, f"//vm[@id = {sos_vm_id + post_vm_id}]")
+        vm_launch_etree = eval_xpath(launch_etree, f"//user_vm[@id='{post_vm_id}']")
+        if vm_scenario_etree is None:
+            logging.warning(f"Post-launched VM {post_vm_id} is not specified in the scenario XML and has no launch script generated.")
```

Comment:
I think we refer to VMs by their name and not their ID now, right?
```suggestion
            logging.warning(f"Post-launched VM {post_vm_id} is not specified in the scenario XML, so no launch script will be generated.")
```

Reply:
Today definitions of the VM in scenario & launch XMLs are linked by the ID, which is why I keep the logic as it was. After we combine those two kinds of XMLs in a week or two, such checks will no longer be needed any more.